### DISCUSSION TITLE
Targeting unknown systems/targets.

### DISCUSSION THREAD

**Discussion by Doubl3MD** [halide/Halide#7640] [Q&A] [answered]
Hi fellow Halidians,

we have a problem to tackle hopefully it's clear what I want to do.

Our context: we are developing an Adobe illustrator plugin. For some compute intensive features we decided to use Halide. It is magicly fast and furious, making the impossible, possible. 

Our problem: when deploying our build plugin (.aip file) from the dev system(win10) to another system (win10 or 11) the halide parts refuse duty. As for now it looks like pipes with parallel and vectorizing are most problematic. 

Our (not a) solution(for now): We rebuild the halide pipes on the other systems and after a rebuild of the Illustrator Plugin everything works great again... but you can see the real bottleneck here... when the day comes to release our plugin to 'unknown' targets/systems this workflow is obviously not the way to go as this is 1: practically impossible and 2: it's just not known what kind of system the client will run his Illustrator on.

The question: what is the typical kind of workflow or build setup, so we can counter this problem of unknown client systems and targets?

Do we build several version of the halide pipes? Run a check on the client system and run the right pipeline? Dono...

Our project really will stand or fall depending on how Halide is flexible in this way.

Any hint or tip or whatever to get us going would be more than welcome.

Thx in advance!

Doubl3MD

**Comment by Doubl3MD** [comment]
Update:

I've suspected the vectorisation features.
So I made a quick tool to compare which features were on or off. On my dev system and another system.
Turns out the other system doesn't have AVX-512 features and my system does. 
So I've disabled the AVX-512 features and now it works.
Being glad with this first victory, my question still stands but it shifted more to this:

Are there pitfalls or others things I should keep in mind to make sure our Illustrator Plugin keeps working on unknown systems/targets?
Does a multi feature target 'waterfall' back on the best feature?
Or can I pick the features to use at runtime when the pipe has been build?

Thx again advance for extra info on this topic!

**Comment by ashishUthama** [comment, marked as answer]
Search for 'multi target' with Halide for more info on this. Halide supports runtime dynamic dispatch based on target capability (in this case SIMD family). 

In short:
1. Make your code SIMD width reactive by using [native_vector_size](https://halide-lang.org/docs/struct_halide_1_1_internal_1_1_schedule_features.html#a56bfc693166da69ca272933bd0e1f1be)
2. At generation time, specify multi-targets like so:   "linux-x86-64-sse41-avx,linux-x86-64-sse41,linux-x86-64". At runtime this causes the code to search in the order specified (left to right), stopping at the first subtarget that is determined to be safe to use.
3. Create the runtime separately ensuring to pick the common denominator

**Reply by steven-johnson** [reply]
This is the right answer :-)

**Comment by Doubl3MD** [comment]
Although that generator system I'm not sure how to use it in windows... This information helped a lot already. Thx!
